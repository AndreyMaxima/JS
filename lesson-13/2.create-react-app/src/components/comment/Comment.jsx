import React from 'react'
import './Comment.css'


export class Comment extends React.Component {
  render(){
    return <div className="comment">
      {console.log('I`m alive') /*Этот код будет выполнен*/}
      <div className="comment__user-name">{this.props.name/*Все переданные пропсы находятся в this.props*/}</div>
      <div className="comment__text">{this.props.text}</div>
    </div>
  }
}

