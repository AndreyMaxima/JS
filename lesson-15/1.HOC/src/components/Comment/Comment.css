.comment {
    border: 1px solid black;
    border-radius: 10px;
    padding: 10px;
}

.comment__user-name {
    font-size: 14px;
    text-align: left;
}

.comment__like {
    text-align: right;
    cursor: pointer;
}

.text-liked {
    color: coral;
}