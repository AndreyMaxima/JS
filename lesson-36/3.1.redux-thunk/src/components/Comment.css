.comment {
    display: flex;
    border: 1px solid gray;
    border-radius: 5px;
}
