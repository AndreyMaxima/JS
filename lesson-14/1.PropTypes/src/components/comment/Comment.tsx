import React, { RefObject } from 'react';
import './Comment.css';

// import PropTypes from 'prop-types'

interface Props { // Объявление интерфейса пропсов
  name?: string;
  text?: string;
  addLike?: () => void;
  removeLike?: () => void;
}

interface State {
  liked: boolean
}

export default class Comment extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props); // Обязательно
    this.handleLike = this.handleLike.bind(this); // Обязательно привязываем this к методам
    this.state = { liked: false };
    this.textDiv = React.createRef(); // Создание ссылки
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  handleLike(e: any) {
    // console.log(e); // Генерируется объект SyntheticEvent (обёртка над event'oм)
    // @ts-ignore
    // this.state.liked = !this.state.liked; Так не работает, стэйт на прямую не меняем
    // alert(this.textDiv.current.innerHTML);
    !this.state.liked
      ? this.props.addLike && this.props.addLike()
      : this.props.removeLike && this.props.removeLike();
    this.setState({ // Так меняем стэйт (планируется смена состояния, происходит не моментально) (изменение внутри рендера приведёт к ошибке)
      liked: !this.state.liked, // Изменится только то, что передали
    });
  }

  textDiv: RefObject<HTMLDivElement>; // Ссылка на элемент

  render() {
    return (
      <div className="comment">
        <div className="comment__user-name">{this.props.name || 'Незнакомец'}</div>
        <div ref={this.textDiv /* присвоение ссылки */} className={`comment__text ${this.state.liked && 'text-liked'}`}>{this.props.text || '-'}</div>
        <div className="comment__like" onClick={this.handleLike}>
          {this.state.liked ? 'Лайкнуто' : 'Не лайкнуто'}
        </div>
      </div>
    );
  }
}

// Comment.propTypes = {
//   name: PropTypes.string, // Назначение типа пропсу
//   text: PropTypes.string,
//   array: PropTypes.array,
//   arrayType: PropTypes.arrayOf(PropTypes.string), // Типизированный массив
//   bool: PropTypes.bool,
//   func: PropTypes.func,
//   num: PropTypes.number,
//   obj: PropTypes.object,
//   reactElem: PropTypes.element, //Экземпляр React-элемента
//   reactClass: PropTypes.elementType, //Экземпляр React-элемента
//   inst: PropTypes.instanceOf(Comment), // Экземпляр конкретного класса,
//   optionalEnum: PropTypes.oneOf(['one', 'two']), // Один из вариантов
//   optionalUnion: PropTypes.oneOfType([PropTypes.string, PropTypes.number]), // Один из типов
//   struct: PropTypes.shape({
//     name: PropTypes.string,
//     age: PropTypes.number
//   })
// }
