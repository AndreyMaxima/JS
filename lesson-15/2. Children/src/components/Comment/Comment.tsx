import React, { RefObject } from 'react';
import './Comment.css';

interface Props {
  name?: string;
  text?: string;
  addLike?: () => void;
  removeLike?: () => void;
}

interface State {
  liked: boolean
}

export default class Comment extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.handleLike = this.handleLike.bind(this);
    this.state = { liked: false };
    this.textDiv = React.createRef();
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  handleLike(e: any) {
    !this.state.liked
      ? this.props.addLike && this.props.addLike()
      : this.props.removeLike && this.props.removeLike();
    this.setState({
      liked: !this.state.liked,
    });
  }

  textDiv: RefObject<HTMLDivElement>;

  render() {
    return (
      <div className="comment">
        <div className="comment__user-name">{this.props.name || 'Незнакомец'}</div>
        <div
          ref={this.textDiv}
          className={`comment__text ${this.state.liked && 'text-liked'}`}
        >
          {this.props.text || '-'}
        </div>
        <div className="comment__like" onClick={this.handleLike}>
          {this.state.liked ? 'Лайкнуто' : 'Не лайкнуто'}
        </div>
      </div>
    );
  }
}
