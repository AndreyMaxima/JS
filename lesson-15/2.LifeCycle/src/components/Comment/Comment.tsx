import React, { RefObject } from 'react';
import './Comment.css';
import LikeButton from '../LikeButton/LikeButton';

interface Props {
  name?: string;
  text?: string;
  addLike?: () => void;
  removeLike?: () => void;
}

interface State {
  liked: boolean
}

export default class Comment extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.handleLike = this.handleLike.bind(this);
    this.state = { liked: false };
    this.textDiv = React.createRef();
  }

  componentWillUnmount(): void {
    console.log('Комментарий размонтирован');
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  handleLike(e: any) {
    !this.state.liked
      ? this.props.addLike && this.props.addLike()
      : this.props.removeLike && this.props.removeLike();
    this.setState({
      liked: !this.state.liked,
    });
  }

  textDiv: RefObject<HTMLDivElement>;

  render() {
    return (
      <div className="comment">
        <div className="comment__user-name">{this.props.name || 'Незнакомец'}</div>
        <div
          ref={this.textDiv}
          className={`comment__text ${this.state.liked && 'text-liked'}`}
        >
          {this.props.text || '-'}
        </div>
        <div className="comment__like">
          <LikeButton liked={this.state.liked} setLiked={this.handleLike} />
        </div>
      </div>
    );
  }
}
