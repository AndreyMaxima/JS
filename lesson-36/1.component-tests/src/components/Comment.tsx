import React, { useState } from 'react';
import './Comment.css';

interface Props {
  message?: string;
  firstName?: string;
  lastName?: string;
  avatarUrl?: string;
}

const Comment = (props: Props) => {
  const {
    message, firstName, lastName, avatarUrl,
  } = props;

  const [liked, setLiked] = useState(false);

  const handleLike = () => {
    setLiked(!liked);
  };

  return (
    <div className="comment">
      <div className="comment_avatar"><img alt="avatar" src={avatarUrl} /></div>
      <div className="comment_content">
        <div className="comment_user-name">{`${firstName || ''} ${lastName || ''}`}</div>
        <div className="comment_message">{message || ''}</div>
      </div>
      <div className="comment_like" onClick={handleLike}>{liked ? 'лайкнуто' : 'не лайкнуто'}</div>
    </div>
  );
};

export default Comment;
