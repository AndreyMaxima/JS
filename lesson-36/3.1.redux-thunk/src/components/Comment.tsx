import React from 'react';
import './Comment.css';

interface Props {
  message?: string;
  firstName?: string;
  lastName?: string;
  avatarUrl?: string;
}

const Comment = (props: Props) => {
  const {
    message, firstName, lastName, avatarUrl,
  } = props;
  return (
    <div className="comment">
      <div className="comment__avatar"><img alt="avatar" src={avatarUrl} /></div>
      <div className="comment__content">
        <div className="comment__user-name">{`${firstName} ${lastName}`}</div>
        <div className="comment__message">{message}</div>
      </div>
    </div>
  );
};

export default Comment;
