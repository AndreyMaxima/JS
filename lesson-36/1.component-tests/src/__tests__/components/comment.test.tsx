import React from 'react';
import { shallow } from 'enzyme';
import Comment from '../../components/Comment';

describe('Comment component test', () => {
  test('should render comment', () => {
    const wrapper = shallow(<Comment />);
    expect(wrapper.find('div.comment')).toHaveLength(1);
  });

  test('should render info', () => {
    const firstName = 'Andreas';
    const lastName = 'Valent';
    const avatar = 'anyUrl';
    const message = 'any text';
    const wrapper = shallow(<Comment
      firstName={firstName}
      lastName={lastName}
      avatarUrl={avatar}
      message={message}
    />);
    expect(wrapper.find('div.comment_avatar').find('img').prop('src')).toBe(avatar);
    expect(wrapper.find('div.comment_user-name').text()).toBe(`${firstName} ${lastName}`);
    expect(wrapper.find('div.comment_message').text()).toBe(message);
    expect(wrapper.find('div.comment_message').text()).toBe(message);
    expect(wrapper.find('div.comment_like').text()).toBe('не лайкнуто');
  });

  test('should handle like', () => {
    const wrapper = shallow(<Comment
      firstName="firstName"
      lastName="lastName"
      avatarUrl="avatar"
      message="message"
    />);
    expect(wrapper.find('div.comment_like').text()).toBe('не лайкнуто');
    wrapper.find('div.comment_like').simulate('click');
    expect(wrapper.find('div.comment_like').text()).toBe('лайкнуто');
    wrapper.find('div.comment_like').simulate('click');
    expect(wrapper.find('div.comment_like').text()).toBe('не лайкнуто');
  });
});

describe('Comment snapshot test', () => {
  test('should render empty comment', () => {
    const wrapper = shallow(<Comment />);
    expect(wrapper).toMatchSnapshot();
  });

  test('should handle like', () => {
    const wrapper = shallow(<Comment />);
    wrapper.find('div.comment_like').simulate('click');
    expect(wrapper).toMatchSnapshot();
    wrapper.find('div.comment_like').simulate('click');
    expect(wrapper).toMatchSnapshot();
  });
});
