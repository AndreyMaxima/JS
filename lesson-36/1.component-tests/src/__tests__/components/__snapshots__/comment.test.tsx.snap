// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment snapshot test should handle like 1`] = `
<div
  className="comment"
>
  <div
    className="comment_avatar"
  >
    <img
      alt="avatar"
    />
  </div>
  <div
    className="comment_content"
  >
    <div
      className="comment_user-name"
    >
       
    </div>
    <div
      className="comment_message"
    />
  </div>
  <div
    className="comment_like"
    onClick={[Function]}
  >
    лайкнуто
  </div>
</div>
`;

exports[`Comment snapshot test should handle like 2`] = `
<div
  className="comment"
>
  <div
    className="comment_avatar"
  >
    <img
      alt="avatar"
    />
  </div>
  <div
    className="comment_content"
  >
    <div
      className="comment_user-name"
    >
       
    </div>
    <div
      className="comment_message"
    />
  </div>
  <div
    className="comment_like"
    onClick={[Function]}
  >
    не лайкнуто
  </div>
</div>
`;

exports[`Comment snapshot test should render empty comment 1`] = `
<div
  className="comment"
>
  <div
    className="comment_avatar"
  >
    <img
      alt="avatar"
    />
  </div>
  <div
    className="comment_content"
  >
    <div
      className="comment_user-name"
    >
       
    </div>
    <div
      className="comment_message"
    />
  </div>
  <div
    className="comment_like"
    onClick={[Function]}
  >
    не лайкнуто
  </div>
</div>
`;
